In [182]:
import pandas as pd

In [183]:
df = pd.read_csv('./heart.csv')

In [184]:
#viewing the dataset too see what we are working with
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [185]:
#from our dataset on column HeartDieses 0 represents no heart dieses and 1 represents heart dieseas 

In [186]:
#checking if the are null values in our dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


In [187]:
#there are no null values in our dataset
#we have 6int columns ,1 float column and 5 object columns

In [188]:
#checking the ranges of the max and min
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  136.809368   
std      9.432617   18.514154   109.384145    0.423046   25.460334   
min     28.000000    0.000000     0.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

                                         Features
Age: age of the patient [years]

Sex: sex of the patient [M: Male, F: Female]

ChestPainType: chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic]

RestingBP: resting blood pressure [mm Hg]

Cholesterol: serum cholesterol [mg/dl]

FastingBS: fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]

RestingECG: resting electrocardiogram results [Normal: Normal, ST: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV), LVH: showing probable or definite left ventricular hypertrophy by Estes' criteria]

MaxHR: maximum heart rate achieved [Numeric value between 60 and 202]

ExerciseAngina: exercise-induced angina [Y: Yes, N: No]

Oldpeak: oldpeak = ST [Numeric value measured in depression]

ST_Slope: the slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]

HeartDisease: output class [1: heart disease, 0: Normal]

In [189]:
#from our data it shows that there are large ranges hence we will have to scale our data
'''also RestingBP,Cholestrol cannot be zero meaning to say either data was not collected or its was only collected when 
one was not feeling so well so in our case we will have to replace those values with the normal values'''

'also RestingBP,Cholestrol cannot be zero meaning to say either data was not collected or its was only collected when \none was not feeling so well so in our case we will have to replace those values with the normal values'

In [190]:
#restingBp==60-100 and cholestrol_when_normal ==125mg/dl-200mg/dl
df['RestingBP'] = df['RestingBP'].replace([0],70)
df['Cholesterol'] = df['Cholesterol'].replace([0],125)

In [191]:
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.472767   222.220044    0.233115  136.809368   
std      9.432617   18.107965    70.883070    0.423046   25.460334   
min     28.000000   70.000000    85.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

In [192]:
#checking the unique values for the object columns
df['ST_Slope'].unique()

array(['Up', 'Flat', 'Down'], dtype=object)

In [193]:
df['ChestPainType'].unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [194]:
df['RestingECG'].unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [195]:
df['ExerciseAngina'].unique()

array(['N', 'Y'], dtype=object)

In [196]:
#As we can see now the min values have been updated now lets encode our object columns
from sklearn.preprocessing import OrdinalEncoder
s = (df.dtypes == 'object')
object_cols = list(s[s].index)
ordinal_encoder = OrdinalEncoder()
df[object_cols] = ordinal_encoder.fit_transform(df[object_cols])

In [197]:
df.head()

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40  1.0            1.0        140          289          0         1.0   
1   49  0.0            2.0        160          180          0         1.0   
2   37  1.0            1.0        130          283          0         2.0   
3   48  0.0            0.0        138          214          0         1.0   
4   54  1.0            2.0        150          195          0         1.0   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172             0.0      0.0       2.0             0  
1    156             0.0      1.0       1.0             1  
2     98             0.0      0.0       2.0             0  
3    108             1.0      1.5       1.0             1  
4    122             0.0      0.0       2.0             0

In [198]:
#from the above columns on Sex-1=Male and 0=Female, RestungECG 1.0 = Normal and 2.0=
df['ChestPainType'].unique()

array([1., 2., 0., 3.])

In [199]:
df['RestingECG'].unique()

array([1., 2., 0.])

In [200]:
df['ExerciseAngina'].unique()

array([0., 1.])

In [201]:
df['ST_Slope'].unique()

array([2., 1., 0.])

From The above the encoded values correspond to the following names

1.ChestPainType:
'ATA', 'NAP', 'ASY', 'TA'==1,2,0,3 respectively

2.RestingECG:
'Normal', 'ST', 'LVH' == 1, 2, 0 respectively

3.ExerciseAngina:
'N', 'Y' == 0, 1 respectively

4.ST_Slope:
'Up', 'Flat', 'Down' == 2, 1, 0 respectively

In [202]:
#checking the correlation in respect to HeartDiesease column
corr_test = df.copy()
correlation_matrix = corr_test.corr()
correlation_matrix['HeartDisease'].sort_values(ascending=False)

HeartDisease      1.000000
ExerciseAngina    0.494282
Oldpeak           0.403951
Sex               0.305445
Age               0.282039
FastingBS         0.267291
RestingBP         0.113788
RestingECG        0.057384
Cholesterol      -0.139484
ChestPainType    -0.386828
MaxHR            -0.400421
ST_Slope         -0.558771
Name: HeartDisease, dtype: float64

In [203]:
corr_chosen = corr_test
correlation_matrixx = corr_chosen.corr()
correlation_matrixx.style.background_gradient(cmap='coolwarm')

From our correlation we can deduce that ExerciseAngina, Oldpeak are strongly +ve correlated to  HeartDiesease and ST_Slope is -ve strongly correlated as well with sex showing the least correlation to heart diesease

In [204]:
#creating our model
y = df['HeartDisease']
X = df.drop('HeartDisease',axis=1)

In [205]:
#scaling X
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

In [206]:
from sklearn.model_selection import train_test_split
train_X,test_X,train_y,test_y = train_test_split(X_scaled,y,test_size=0.2,random_state=42)
train_y

795    0
25     0
84     1
10     0
344    1
      ..
106    0
270    0
860    1
435    0
102    1
Name: HeartDisease, Length: 734, dtype: int64

In [207]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
mod = RandomForestRegressor(n_estimators=1)
mod.fit(train_X, train_y)
preds = mod.predict(test_X)
print(mean_absolute_error(test_y, preds))

0.19021739130434784


In [208]:
#our mean absolute error is not that bad but there is always room for fine tuning if one feels like fine tuning the model

In [210]:
#dumping the model
import pickle

pickle.dump(mod, open('model.pkl', 'wb'))